File: go/questions/question_three.go
package questions

import (
	"strconv"
	"unicode"
)

func RunLengthEncode(s string) string {
	encodedString := ""
	count := 1
	for i := 0; i < len(s); i++ {
		str := string(s[i])
		if i < len(s)-1 && s[i] == s[i+1] {
			count++
		} else {
			encodedString += str + strconv.Itoa(count)
			count = 1
		}
	}
	return encodedString
}

func RunLengthDecode(s string) string {
	decodedString := ""
	countString := ""
	left := 0
	right := 1

	for right < len(s) {
		if unicode.IsLetter(rune(s[right])) == false {
			countString += string(s[right])
			right++
		}
		if right == len(s) || unicode.IsLetter(rune(s[right])) {
			var count int
			if countString == "" {
				count = 1
			} else {
				count, _ = strconv.Atoi(countString)
			}
			decodedString = populateString(decodedString, string(s[left]), count)
			countString = ""
			left = right
			right++
		}
	}
	return decodedString
}

func populateString(stringToPopulate string, repeatedChar string, count int) string {
	for i := 0; i < count; i++ {
		stringToPopulate += repeatedChar
	}
	return stringToPopulate
}

File: java/src/main/java/QuestionThree.java
public class QuestionThree {

	public static String runLengthEncode(String s) {
		String encodedString = "";
		int count = 1;
		for (int i = 0; i < s.length(); i++) {
			if (i < s.length() - 1 && s.charAt(i) == s.charAt(i + 1)) {
				count++;
			} else {
				encodedString += s.charAt(i) + Integer.toString(count);
				count = 1;
			}
		}
		return encodedString;
	}

	public static String runLengthDecode(String s) {
		String decodedString = "";
		String countString = "";
		int left = 0;
		int right = 1;

		while (right < s.length()) {
			if (!Character.isAlphabetic(s.charAt(right))) {
				countString += s.charAt(right);
				right++;
			}
			if (right == s.length() || Character.isAlphabetic(s.charAt(right))) {
				int count;
				if (countString == "") {
					count = 1;
				} else {
					count = Integer.parseInt(countString);
				}
				decodedString = populateString(decodedString, s.charAt(left), count);
				countString = "";
				left = right;
				right++;
			}
		}
		return decodedString;
	}

	private static String populateString(String stringToPopulate, char c, int count) {
		for (int j = 0; j < count; j++) {
			stringToPopulate += c;
		}
		return stringToPopulate;
	}
}
